ambry-store/src/main/java/com.github.ambry.store/ScanResults.java

```diff
@@ -0,0 +1,213 @@
+      containerBucketTimeMs += bucketSpanInMs;
+      logSegmentBucketTimeMs += bucketSpanInMs;
+    }
+    containerForecastStartTimeMs = containerBuckets.firstKey();
+    containerLastBucketTimeMs = containerBuckets.lastKey();
+    containerForecastEndTimeMs = containerLastBucketTimeMs + bucketSpanInMs;
+    logSegmentForecastStartTimeMs = logSegmentBuckets.firstKey();
+    logSegmentLastBucketTimeMs = logSegmentBuckets.lastKey();
+    logSegmentForecastEndTimeMs = logSegmentLastBucketTimeMs + bucketSpanInMs;
+  }
+
+  /**
+   * Given a reference time, return the key of the appropriate container bucket whose end time is strictly greater than
+   * the reference time.
+   * @param referenceTimeInMs the reference time or operation time of an event.
+   * @return the appropriate bucket key (bucket end time) to indicate which bucket will an event with
+   * the given reference time as operation time belong to.
+   */
+  Long getContainerBucketKey(long referenceTimeInMs) {
+    return containerBuckets.higherKey(referenceTimeInMs);
+  }
+
+  /**
+   * Given a reference time, return the key of the appropriate log segment bucket whose end time is strictly greater
+   * than the reference time.
+   * @param referenceTimeInMs the reference time or operation time of an event.
+   * @return the appropriate bucket key (bucket end time) to indicate which bucket will an event with
+   * the given reference time as operation time belong to.
+   */
+  Long getLogSegmentBucketKey(long referenceTimeInMs) {
+    return logSegmentBuckets.higherKey(referenceTimeInMs);
+  }
+
+  /**
+   * Helper function to update the container base value bucket with the given value.
+   * @param serviceId the serviceId of the map entry to be updated
+   * @param containerId the containerId of the map entry to be updated
+   * @param value the value to be added
+   */
+  void updateContainerBaseBucket(String serviceId, String containerId, long value) {
+    updateContainerBucket(containerBuckets.firstKey(), serviceId, containerId, value);
+  }
+
+  /**
+   * Helper function to update the log segment base value bucket with the given value.
+   * @param logSegmentName the log segment name of the map entry to be updated
+   * @param value the value to be added
+   */
+  void updateLogSegmentBaseBucket(String logSegmentName, long value) {
+    updateLogSegmentBucket(logSegmentBuckets.firstKey(), logSegmentName, value);
+  }
+
+  /**
+   * Helper function to update a container bucket with the given value.
+   * @param bucketKey the bucket key to specify which bucket will be updated
+   * @param serviceId the serviceId of the map entry to be updated
+   * @param containerId the containerId of the map entry to be updated
+   * @param value the value to be added
+   */
+  void updateContainerBucket(Long bucketKey, String serviceId, String containerId, long value) {
+    if (bucketKey != null && containerBuckets.containsKey(bucketKey)) {
```

Comment:
passing a null key is not legal. It would only happen if containerBuckets.higherKey() returns null for a valid reference time (not possible since within range is checked before calling getContainerBucketKey).